pkg/visualize/graphviz.go

```diff
@@ -0,0 +1,361 @@
+package visualize
+
+import (
+	"bytes"
+	"context"
+	"fmt"
+	"strings"
+
+	"github.com/flyteorg/flyteidl/clients/go/coreutils"
+
+	"github.com/flyteorg/flyteidl/gen/pb-go/flyteidl/core"
+	"github.com/flyteorg/flytestdlib/errors"
+	"github.com/flyteorg/flytestdlib/logger"
+	"github.com/goccy/go-graphviz"
+	"github.com/goccy/go-graphviz/cgraph"
+)
+
+func operandToString(op *core.Operand) string {
+	if op.GetPrimitive() != nil {
+		l, err := coreutils.ExtractFromLiteral(&core.Literal{Value: &core.Literal_Scalar{
+			Scalar: &core.Scalar{
+				Value: &core.Scalar_Primitive{
+					Primitive: op.GetPrimitive(),
+				},
+			},
+		}})
+		if err != nil {
+			return err.Error()
+		}
+		return fmt.Sprintf("%v", l)
+	}
+	return op.GetVar()
+}
+
+func comparisonToString(expr *core.ComparisonExpression) string {
+	return fmt.Sprintf("%s %s %s", operandToString(expr.LeftValue), expr.Operator.String(), operandToString(expr.RightValue))
+}
+
+func conjunctionToString(expr *core.ConjunctionExpression) string {
+	return fmt.Sprintf("(%s) %s (%s)", booleanExprToString(expr.LeftExpression), expr.Operator.String(), booleanExprToString(expr.RightExpression))
+}
+
+func booleanExprToString(expr *core.BooleanExpression) string {
+	if expr.GetConjunction() != nil {
+		return conjunctionToString(expr.GetConjunction())
+	}
+	return comparisonToString(expr.GetComparison())
+}
+
+func constructStartNode(n string, graph *cgraph.Graph) (*cgraph.Node, error) {
+	gn, err := graph.CreateNode(n)
+	if err != nil {
+		return nil, err
+	}
+	gn.SetLabel("start")
+	gn.SetShape(cgraph.DoubleCircleShape)
+	gn.SetColor("green")
+	return gn, nil
+}
+
+func constructEndNode(n string, graph *cgraph.Graph) (*cgraph.Node, error) {
+	gn, err := graph.CreateNode(n)
+	if err != nil {
+		return nil, err
+	}
+	gn.SetLabel("end")
+	gn.SetShape(cgraph.DoubleCircleShape)
+	gn.SetColor("red")
+	return gn, nil
+}
+
+func constructTaskNode(name string, graph *cgraph.Graph, n *core.Node, t *core.CompiledTask) (*cgraph.Node, error) {
+	gn, err := graph.CreateNode(name)
+	if err != nil {
+		return nil, err
+	}
+	if n.Metadata != nil && n.Metadata.Name != "" {
+		v := strings.LastIndexAny(n.Metadata.Name, ".")
+		gn.SetLabel(fmt.Sprintf("%s [%s]", n.Metadata.Name[v+1:], t.Template.Type))
```

Comment:
nit: in the examples it looks like the task type always goes outside the bounding box for the task node - is this fixable here?

Reply:
hmm, fixable probably by using a better way of rendering

Reply:
💯 sounds good!